src/sensors/distant.cpp

```diff
@@ -0,0 +1,494 @@
+     with zero weights. *Note*: ray invalidation occurs per-lane in packet
+     modes.
+
+This sensor plugin implements a distant directional sensor which records
+radiation leaving the scene in a given direction. It records the spectral
+radiance leaving the scene in the specified direction.
+Rays target points are sampled from the cross section of the scene's bounding
+sphere and their origins are positioned outside of the scene's geometry.
+
+Based on the film size, the ray direction sampling strategy will vary:
+
+* if a 1x1 film is passed, rays directions will be equal to ``-direction``
+  (unless ``flip_directions`` is true, in which case they will be equal to
+  ``direction``);
+* if a Nx1 film is passed (*i.e.* if film height is reduced to 1), ray
+  directions will be sampled from the intersection of the hemisphere defined by
+  ``-direction`` and the (vector) plane generated by ``orientation`` and
+  ``direction``;
+* if a NxM film is passed, ray directions will be sampled in the hemisphere
+  defined by ``-direction``.
+
+Rays sampled from this sensor can be tuned so as to target a specific region of
+the scene using the ``ray_target`` parameter. The recorded radiance is averaged
+over the targeted geometry.
+
+The positioning of the origin of those rays can also be controlled using the
+``ray_origin``. This is particularly useful when the scene has a dimension much
+smaller than the others and it is not necessary that ray origins are located at
+the scene's bounding sphere.
+
+.. warning::
+
+   If this sensor is used with a targeting strategy leading to rays not hitting
+   the scene's geometry (*e.g.* default targeting strategy), it will pick up
+   ambiant emitter radiance samples (or zero values if no ambiant emitter is
+   defined). Therefore, it is almost always preferrable to use a nondefault
+   targeting strategy.
+*/
+
+template <typename Float, typename Spectrum>
+class DistantSensor final : public Sensor<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(Sensor, m_world_transform, m_film)
+    MTS_IMPORT_TYPES(Shape)
+
+    DistantSensor(const Properties &props) : Base(props), m_props(props) {
+
+        // Get target
+        if (props.has_property("ray_target")) {
+            try {
+                // We first try to get a point
+                props.point3f("ray_target");
+                m_ray_target_type = RayTargetType::Point;
+            } catch (const std::runtime_error &e) {
+                // If it fails, we assume it's a shape
+                m_ray_target_type = RayTargetType::Shape;
+            }
+        } else {
+            m_ray_target_type = RayTargetType::None;
+        }
+
+        // Get origin
+        if (props.has_property("ray_origin"))
+            m_ray_origin_type = RayOriginType::Shape;
+        else {
+            m_ray_origin_type = RayOriginType::BoundingSphere;
+        }
+
+        props.mark_queried("direction");
+        props.mark_queried("flip_directions");
```

Comment:
What would be a use case for the `flip_directions` parameter?

Reply:
When using the Dirac direction sampling strategy (1x1 film), the sensor looks, by default, in the direction opposite to the `direction` parameter. This is unintuitive, especially because the `directional` plugin uses a different convention. See a more general comment on this below.